Upstream patch to implement follow remain strategy
Respective changelogs:
2015-08-05  Gavin Smith  <gavinsmith0123@gmail.com>
 
	* info/variables.c (info_variables): New user variable 
	"follow-strategy".
	* info/nodes.c (info_get_node_with_defaults): If 
	follow-strategy=remain (the default), try to remain within the 
	same directory as the node we're coming from.

To disable it, set the "follow-strategy" variable to "path" in .infokey.
---
 info/nodes.c     |   36 ++++++++++++++++++++++++++++++++++--
 info/variables.c |    6 ++++++
 info/variables.h |    1 +
 3 files changed, 41 insertions(+), 2 deletions(-)

--- texinfo.orig/info/nodes.c
+++ texinfo/info/nodes.c
@@ -895,6 +895,11 @@
   return i - binding->start;
 }
 
+#define FOLLOW_REMAIN 0
+#define FOLLOW_PATH 1
+
+int follow_strategy;
+
 /* Return a pointer to a NODE structure for the Info node (FILENAME)NODENAME,
    using DEFAULTS for defaults.  If DEFAULTS is null, the defaults are:
    - If FILENAME is NULL, `dir' is used.
@@ -928,8 +933,35 @@
       goto cleanup_and_exit;
     }
 
-  /* Find the correct info file, or give up.  */
-  file_buffer = info_find_file (filename);
+
+  if (follow_strategy == FOLLOW_REMAIN
+      && defaults && defaults->fullpath)
+    {
+      /* Find the directory in the filename for defaults, and look in
+         that directory first. */
+      char *file_in_same_dir;
+      char saved_char, *p;
+
+      p = defaults->fullpath + strlen (defaults->fullpath);
+      while (p > defaults->fullpath && !IS_SLASH (*p))
+        p--;
+
+      if (p > defaults->fullpath)
+        {
+          saved_char = *p;
+          *p = 0;
+
+          file_in_same_dir = info_add_extension (defaults->fullpath,
+                                                 filename, 0);
+          if (file_in_same_dir)
+            file_buffer = info_find_file (file_in_same_dir);
+          free (file_in_same_dir);
+          *p = saved_char;
+        }
+    }
+
+  if (!file_buffer)
+    file_buffer = info_find_file (filename);
 
   if (file_buffer)
     {
--- texinfo.orig/info/variables.c
+++ texinfo/info/variables.c
@@ -36,6 +36,8 @@
 
 static char *mouse_choices[] = { "Off", "normal-tracking", NULL };
 
+static char *follow_strategy_choices[] = { "remain", "path", NULL };
+
 /* Choices used by the completer when reading a value for the user-visible
    variable "scroll-behaviour". */
 static char *info_scroll_choices[] = { "Continuous", "Next Only",
@@ -124,6 +126,10 @@
       N_("Method to use to track mouse events"),
     &mouse_protocol, (char **)mouse_choices },
 
+  { "follow-strategy",
+      N_("Method to use to track mouse events"),
+    &follow_strategy, (char **)follow_strategy_choices },
+
   { NULL }
 };
 
--- texinfo.orig/info/variables.h
+++ texinfo/info/variables.h
@@ -82,6 +82,7 @@
 extern int key_time;
 extern int highlight_searches_p;
 extern int mouse_protocol;
+extern int follow_strategy;
 
 
 #endif /* not INFO_VARIABLES_H */
